### PR TITLE
Provide REPORT method on files DAV to filter/search files

### PR DESCRIPTION
**Pull request by PVince81**
Makes it possible to search files by system tags (only, for now).

Steps:
1. Create a report xml file for the query body: "report-systemtags.xml"

``` xml
<?xml version="1.0" encoding="utf-8" ?>
<oc:filter-files xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns" >
        <a:prop>
                <oc:id/>
                <oc:fileid/>
                <oc:permissions/>
                <a:getlastmodified/>
                <a:getetag/>
                <oc:downloadURL/>
                <oc:ddC/>
                <oc:size/>
                <oc:owner-id/>
                <oc:owner-display-name/>
                <oc:size/>
                <oc:checksum />
                <oc:tags />
                <a:quota-used-bytes/>
                <a:quota-available-bytes/>
        </a:prop>
        <oc:filter-rules>
                <oc:systemtag>1</oc:systemtag>
        </oc:filter-rules>
</oc:filter-files>
```
1. In the web UI, create several files and folders
2. Tag some files and folders (not all) with a new tag "test1" (assuming it will have the id 1, else you'll need to adjust the id in the XML above)
3. Request the REPORT: `curl -i --data-binary "@report-systemtags.xml" -X REPORT -H "Content-Type: text/xml" http://root:admin@localhost/owncloud/remote.php/files/`

The result looks like this for me (contains only the tagged files):

``` xml
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/files/welcome.txt</d:href>
  <d:propstat>
   <d:prop>
    <oc:id>00000003ockbfur1449k</oc:id>
    <oc:fileid>3</oc:fileid>
    <oc:permissions>RDNVW</oc:permissions>
    <d:getlastmodified>Wed, 03 Feb 2016 19:23:35 GMT</d:getlastmodified>
    <d:getetag>&quot;c0741586516b87b952f00d26c844ca92&quot;</d:getetag>
    <oc:downloadURL></oc:downloadURL>
    <oc:owner-id>root</oc:owner-id>
    <oc:owner-display-name>root</oc:owner-display-name>
    <oc:checksum></oc:checksum>
    <oc:tags/>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
  <d:propstat>
   <d:prop>
    <oc:ddC/>
    <oc:size/>
    <d:quota-used-bytes/>
    <d:quota-available-bytes/>
   </d:prop>
   <d:status>HTTP/1.1 404 Not Found</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/files/test</d:href>
  <d:propstat>
   <d:prop>
    <oc:id>00000003ockbfur1449k</oc:id>
    <oc:fileid>3</oc:fileid>
    <oc:permissions>RDNVW</oc:permissions>
    <d:getlastmodified>Wed, 03 Feb 2016 19:23:35 GMT</d:getlastmodified>
    <d:getetag>&quot;c0741586516b87b952f00d26c844ca92&quot;</d:getetag>
    <oc:downloadURL></oc:downloadURL>
    <oc:size>1978913</oc:size>
    <oc:owner-id>root</oc:owner-id>
    <oc:owner-display-name>root</oc:owner-display-name>
    <oc:checksum></oc:checksum>
    <oc:tags/>
    <d:quota-used-bytes>1978913</d:quota-used-bytes>
    <d:quota-available-bytes>1070305280</d:quota-available-bytes>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
  <d:propstat>
   <d:prop>
    <oc:ddC/>
   </d:prop>
   <d:status>HTTP/1.1 404 Not Found</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/files/test/Huoguo.jpg</d:href>
  <d:propstat>
   <d:prop>
    <oc:id>00000003ockbfur1449k</oc:id>
    <oc:fileid>3</oc:fileid>
    <oc:permissions>RDNVW</oc:permissions>
    <d:getlastmodified>Wed, 03 Feb 2016 19:23:35 GMT</d:getlastmodified>
    <d:getetag>&quot;c0741586516b87b952f00d26c844ca92&quot;</d:getetag>
    <oc:downloadURL></oc:downloadURL>
    <oc:size>1978913</oc:size>
    <oc:owner-id>root</oc:owner-id>
    <oc:owner-display-name>root</oc:owner-display-name>
    <oc:checksum></oc:checksum>
    <oc:tags/>
    <d:quota-used-bytes>1978913</d:quota-used-bytes>
    <d:quota-available-bytes>1070305280</d:quota-available-bytes>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
  <d:propstat>
   <d:prop>
    <oc:ddC/>
   </d:prop>
   <d:status>HTTP/1.1 404 Not Found</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>
```

You can even add a subfolder in the curl request, it will narrow down the results by folders :smile: 

Note: the way how filtering is done and gathering the results might be very inefficient.

Please check this out @blizzz @icewind1991 @DeepDiver1975 @evert

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] BUG: resourcetype (when requested) doesn't properly contain "collection" in the response for folders

**Comment by PVince81** [comment]
Also if you add multiple "oc:systemtag" entries, it will fetch all files that have ALL these tags assigned (AND query), but the way it does is horrible: fetch the fileids for every tag, then use array_intersect.

**Comment by PVince81** [comment]
- [ ] TODO: refuse to return ALL files if no filter rule was set

**Comment by PVince81** [comment]
- [ ] REQUIRES: https://github.com/owncloud/core/pull/22200 CC @rullzer

**Comment by PVince81** [comment]
- [ ] TODO: prevent filtering by non-visible tags when not admin

**Comment by PVince81** [comment]
- [ ] finish unit tests

File: apps/dav/lib/connector/sabre/filesreportplugin.php [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+	 * @return bool
+	 * @throws NotFound
+	 * @throws ReportNotSupported
+	 */
+	public function onReport($reportName, $report, $uri) {
+		$reportTargetNode = $this->server->tree->getNodeForPath($uri);
+		if (!$reportTargetNode instanceof Directory || $reportName !== self::REPORT_NAME) {
+			throw new ReportNotSupported();
+		}
+
```
Comment (icewind1991) [inline apps/dav/lib/connector/sabre/filesreportplugin.php]: inject this?

**Comment by PVince81** [comment]
Added tests.

Ready for review @icewind1991 @blizzz @DeepDiver1975 @evert @nickvergessen @rullzer @MorrisJobke 

Test instructions here https://github.com/owncloud/core/pull/22112#issue-131161527
Trying to filter by a non-existing tag or user-invisible tag (as user) will throw precondition failed.

**Comment by karlitschek** [comment]
nice :+1:

**Comment by PVince81** [comment]
If you think that this PR is too boring to test because of the many commands, then just review the frontend PR https://github.com/owncloud/core/issues/21001 which includes the backend commits. The frontend uses this call to filter by tags, so you can see that it works :smile:

**Comment by rullzer** [comment]
Played around with the propfind and it retruned what I expected.
Also played with the frontend in #21001 and that also returned what I expected.
Code looks good. :+1:

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**